### DISCUSSION TITLE
export QueryClientContext

### DISCUSSION THREAD

**Discussion by bertho-zero** [TanStack/query#3075] [Ideas]
Hi,

I would like to do an equivalent of `contextSharing` but without going through `window`, just pick up the parent `queryClient` if it exists but its not possible because of that line:

https://github.com/tannerlinsley/react-query/blob/ebf1e68a5258398ab02e9965c876a4f4dfc6f676/src/react/QueryClientProvider.tsx#L38

```js
import React from 'react';
import {
  QueryClient,
  QueryClientProvider as Provider,
  useQueryClient
} from 'react-query';
import useConstant from '../hooks/useConstant';

export default function QueryClientProvider({
  client: customClient,
  contextSharing,
  children
}) {
  const parentQueryClient = useQueryClient(); // throw without parent provider and no way to use the context

  const value = useConstant(
    () => customClient
      || parentQueryClient
      || new QueryClient({
        defaultOptions: {
          queries: {
            refetchOnWindowFocus: false,
            retry: false,
          },
        },
      })
  );

  return (
    <Provider client={value} contextSharing={contextSharing}>
      {children}
    </Provider>
  );
};

```

I could use the context to look if there is already in a queryClient context without throwing an error but it is not exported.

**Comment by TkDodo** [comment]
would this PR be helpful?

https://github.com/tannerlinsley/react-query/pull/2548

**Reply by bertho-zero** [reply]
It can be a solution but I would prefer not to specify a context every time.

A variant of `useQueryClient` that doesn't throw an error or a way to directly use the context would work fine.

**Reply by TkDodo** [reply]
> A variant of useQueryClient that doesn't throw an error or a way to directly use the context would work fine.

I highly doubt that this is possible. `useQueryClient()` returns a `QueryClient`. If it doesn't throw, it would have to return `QueryClient | undefined`, which would make it unnecessarily hard to work with.

**Reply by bertho-zero** [reply]
How about exporting the default context? it would allow to use the parent context without having any error, without using window and without having to pass a context option everywhere in the project. react-redux exports it for example.

It seems to me useful and the most suitable.
